tik/lib/CartographerKernel.cpp

```diff
@@ -291,500 +281,926 @@ namespace TraceAtlas::tik
-                for (auto bi = cb->begin(); bi != cb->end(); bi++)
+                if (auto inst = dyn_cast<Instruction>(it))
                 {
-                    if (auto *p = dyn_cast<PHINode>(bi))
+                    if (inst->getType()->getTypeID() != Type::VoidTyID)
                     {
-                        for (auto pred : p->blocks())
+                        for (auto use : inst->users())
                         {
-                            if (blocks.find(pred) == blocks.end())
+                            if (auto useInst = dyn_cast<Instruction>(use))
                             {
-                                //we have an invalid predecessor, replace with init
-                                bool found = false;
-                                for (const auto &ent : Entrances)
+                                for (auto ex : exits)
                                 {
-                                    if (IDToBlock[ent->Block] == block)
+                                    if (isPotentiallyReachable(ex, useInst->getParent()))
                                     {
-                                        p->replaceIncomingBlockWith(pred, Init);
-                                        rescheduled++;
-                                        found = true;
-                                        break;
-                                    }
-                                }
-                                if (!found)
-                                {
-                                    auto a = p->getBasicBlockIndex(pred);
-                                    if (a != -1)
-                                    {
-                                        p->removeIncomingValue(pred);
+                                        auto sExtVal = GetValueID(inst);
+                                        if (kernelIE.find(sExtVal) == kernelIE.end())
+                                        {
+                                            KernelExports.push_back(sExtVal);
+                                            kernelIE.insert(sExtVal);
+                                        }
+                                        else if (find(KernelImports.begin(), KernelImports.end(), sExtVal) != KernelImports.end())
+                                        {
+                                            throw AtlasException("Import needs to be an export!");
+                                        }
                                     }
                                 }
                             }
                         }
                     }
-                    else
-                    {
-                        break;
-                    }
-                }
-                if (rescheduled > 1)
-                {
-                    spdlog::warn("Rescheduled more than one phi predecessor"); //basically this is a band aid. Needs some more help
                 }
             }
         }
     }
```

Comment:
I really want some pseudo-code of what this is doing. I'm really struggling to read it due to the diff and do not quite understand what the problem is or what you are doing.

Reply:
The description of GetBoundaryValues() summarizes the core of this routine. It is the naming, gathering of all function arguments and all the metadata (arg names) and error handling (bad args) that go along with that, and a mapping of original values to input args in the VMap.
Name -> function args gathering -> function object construction, arg naming and mapping, error checks -> VMap of the imports